File: 알렉스데브/메뉴리뉴얼/menuRenewal.js
/**공통문자열을 찾는 문제
 *  문자열의 요소로 조합 (combination) 을 만드는 방법: 모든 경우의 수를 따져보는 것이기 때문에 완전탐색이다.
 * 각각의 문자열을 순회하면서 공통문자열의 길이가 2이고 같은 요소가 2번 이상 있을 경우, 새로운 배열에 공통문자열을 추가한다.
 *  중복을 배제하고, 입출력 예의 법칙을 따라 각각으 공통문자열은 사전순으로 정렬하여 추가
 */

function solution(orders, course) {
  let answer = [];
  const single = orders.map((x) => x.split(''));
  console.log(single);
  return answer;
}

console.log(
  solution(['ABCFG', 'AC', 'CDE', 'ACDE', 'BCFG', 'ACDEH'], [2, 3, 4])
);

File: 알렉스데브/메뉴리뉴얼/menurenewal.py
from itertools import combinations

def solution (orders, course):
  answer = []

  #코스요리 단품개수를 순회하며 dict를 선언
  for dish in course:
    dict = {}

    # 주문을 순회하며
    for i in range(len(orders)):

      # 주문에서 메뉴의 조합을 만들고, 그것을 이어 문자열로 만든 것을 요소로 하는 배열 선언.
      comb = [ ''.join(c) for c in combinations(sorted(orders[i]), dish)]

      # 메뉴 조합 배열을 순회하며
      for j in range(len(comb)):

        # 조합 가능한 것을 dict으로 카운트
        dict[comb[j]] = dict.get(comb[j], 0) +1

  # 딕셔너리를 순회하며 dict 안의 최대값을 구하고
  for k in dict:
    DICT_MAX = max(dict.values())
    #딕셔너리의 최대값이 2 이상이고 순회하는 각각 값이 최대값과 같을 경우
    if DICT_MAX >= 2 and dict[k] == DICT_MAX:
      # 조건에 맞으므로 정답 배열에 넣는다.
      answer.append(k)
  return sorted(answer)
  
print(solution(["ABCFG", "AC", "CDE", "ACDE", "BCFG", "ACDEH"], [2,3,4]))